test/0dt/mzcompose.py

```diff
@@ -865,6 +863,343 @@ def workflow_basic(c: Composition) -> None:
         )
 
 
+def workflow_kafka_source_rehydration(c: Composition) -> None:
+    """Verify Kafka source rehydration in 0dt deployment"""
+    c.down(destroy_volumes=True)
+    c.up("zookeeper", "kafka", "schema-registry", "mz_old")
+    c.up("testdrive", persistent=True)
+
+    count = 1000000
+    repeats = 20
+
+    # Make sure cluster is owned by the system so it doesn't get dropped
+    # between testdrive runs.
+    c.sql(
+        """
+        DROP CLUSTER IF EXISTS cluster CASCADE;
+        CREATE CLUSTER cluster SIZE '1';
+        GRANT ALL ON CLUSTER cluster TO materialize;
+        ALTER SYSTEM SET cluster = cluster;
+    """,
+        service="mz_old",
+        port=6877,
+        user="mz_system",
+    )
+
+    start_time = time.time()
+    c.testdrive(
+        dedent(
+            f"""
+        > SET CLUSTER = cluster;
+
+        > CREATE CONNECTION IF NOT EXISTS kafka_conn FOR KAFKA BROKER '${{testdrive.kafka-addr}}', SECURITY PROTOCOL = 'PLAINTEXT';
+        > CREATE CONNECTION IF NOT EXISTS csr_conn FOR CONFLUENT SCHEMA REGISTRY URL '${{testdrive.schema-registry-url}}';
+
+        $ kafka-create-topic topic=kafka-large
+        $ kafka-ingest format=bytes key-format=bytes key-terminator=: topic=kafka-large repeat={count}
+        key0A,key${{kafka-ingest.iteration}}:value0A,${{kafka-ingest.iteration}}
+        > CREATE SOURCE kafka_source
+          IN CLUSTER cluster
+          FROM KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-kafka-large-${{testdrive.seed}}');
+
+        > CREATE TABLE kafka_source_tbl (key1, key2, value1, value2)
+          FROM SOURCE kafka_source (REFERENCE "testdrive-kafka-large-${{testdrive.seed}}")
+          KEY FORMAT CSV WITH 2 COLUMNS DELIMITED BY ','
+          VALUE FORMAT CSV WITH 2 COLUMNS DELIMITED BY ','
+          ENVELOPE UPSERT;
+        # TODO: This should also work with a default index
+        # > CREATE DEFAULT INDEX ON kafka_source_tbl
+        > SELECT COUNT(*) FROM kafka_source_tbl
```

Comment:
The default index on the source isn't the thing that really ought to speed this up—it's indexing the count itself that would do the trick:

```suggestion
        # > CREATE DEFAULT INDEX ON kafka_source_tbl
        > CREATE VIEW kafka_source_cnt AS SELECT count(*) FROM kafka_source_tbl
        > CREATE DEFAULT INDEX on kafka_source_cnt
        > SELECT * FROM kafka_source_cnt
```

Don't need to do this in this PR since I imagine that's broken in the same way that the version you have commented out here, but when we _are_ ready to tackle that, we ought to index the count and not the source directly.